### ISSUE TITLE
Idea: Reimplement timeout as percentage value / float

### ISSUE THREAD

**Issue by JanPietrzyk** [infection/infection#38]
Hi,

I am currently moving an existing project from humbug to infection and have an issue related to the current implementation of the timeout property.

My existing project has a few exceptional tests that take a lot more time then 10 seconds, but most tests are < 1 second of course. Now in order to get infection to execute the test suite initially and generate the coverage  I have to set a high value as timeout. This does only extend the time infection then takes to detect infinite loops in the actual mutation run.

My current solution is to exclude these tests through a custom config. Nevertheless I think there is room for some improvement in infection itself.

My Suggestion:
- Remove the timeout from the initial run alltogether and trust the users here.
- Refactor the timeout into a percentage value. Mark tests that take X% longer then when they were executed initially as timeout.

**Comment by borNfreee** [comment]
Thanks for posting your idea Jan.

> Now in order to get infection to execute the test suite initially and generate the coverage I have to set a high value as timeout. This does only extend the time infection then takes to detect infinite loops in the actual mutation run. 
> [...]
> Remove the timeout from the initial run alltogether and trust the users here.

Let's think together about what the `timeout` property is.

https://infection.github.io/guide/usage.html#Configuration-settings

> `timeout` - the allowed timeout configured for Infection. Make sure to set it to higher value than your tests are executed in seconds to avoid false-positives.

So, if your tests take 1 minutes (does not matter that 1 test takes 59 seconds and all others just one second) - timeout should be at least 1 minute. 

Why? Because infection (and actually humbug too) runs all tests that executed on the mutated line. If you have e.g. 10 tests and all these 10 tests cover all lines, then each mutation process will run all tests, right?

So I don't understand why you want to have `timeout=1s` when your tests (the whole test suite) take much longer - 1+59=60s.

I agree that probably, we need to remove timeout from the initial tests run, but I don't agree that we need to have timeout less in seconds than the whole test suite takes.

Could you please post your thoughts about this?

-----

> Refactor the timeout into a percentage value. Mark tests that take X% longer then when they were executed initially as timeout.

I can suggest having both - timeout float value (current implementation) and a percentage (new functionality). This is exaclty what Pitest (Java mutation testing framework) has: http://pitest.org/quickstart/maven/ (see `timeoutFactor` and `timeoutConstant`)

Do you want to contribute ;) ?

**Comment by JanPietrzyk** [comment]
hmm, this leaves a lot of ideas in the wild ;) Let me try to sort my opinion into this...

> Let's think together about what the timeout property is.

Or more precise why mutation testing needs to have this data. I always thought it is the very simple fact that mutating production code might lead to infinite loops, so the framework needs a stop mark where running a test further is not a valuable option.  (Correct me if I am wrong here)

So If I have a test suite that takes 1 Minute as a whole, consists of 4 tests that are distributed like 40 Sec, 10 Sec, 5 Sec and 5 Sec. What would be a meaningful timeout for the 10 Sec test running against a mutant?

- 10 Sec?
- 1 Minute?
- 10 Sec * 4.50?

This is a problem of relation. The fixed information a mutation testing framework has is in the test and the initial run. Which leaves us with the following data points:

* 1 Min execution time as a whole in the happy case
* 10 Sec test execution time in the happy case

So now the framework mutates the production code and obfuscates the happy cases. The original timing therefore gets obfuscated as well. There is no guarantee that the 1 Minute mark still holds up, as there is no guarantee that the 10 Sec mark still applies. So to me it looks like both are a little arbitrary.

Further more it looks like this is confusing performance testing with mutation testing. The application certainly has to meet requirements in this regard, but not enforced through the mutation testing framework.

Continuing with the 1 Minute example above, infection then has a worst case runtime (single threaded) of 4 Minutes (Number of tests * Timeout), adding in that the test suite has to run on different machines it would be more like 5 Minutes. This leaves me with a key problem: **There is no meaningful upper boundary for runtime.**

As an example: My current project has 1,566 tests and executes in 3 Minutes, which means that infection can run for 78 hours.

> I agree that probably, we need to remove timeout from the initial tests run, but I don't agree that we need to have timeout less in seconds than the whole test suite takes.

So using a lower timeout seems like the pragmatic thing to do if the number of tests exceeds a certain threshold. But having the same timeout in the whole run, makes this currently impossible :(

> Do you want to contribute ;) ?

:) Maybe, I certainly did a lot of digging to find the cause. What are the quality requirements for PR's?

What are your thoughts on this? Is there a more dogmatic solution, that I don't see?

**Comment by borNfreee** [comment]
> so the framework needs a stop mark where running a test further is not a valuable option. (Correct me if I am wrong here)

I should say that `timeout` property is a timeout for the *whole testsuite*, not for the particular test from the suite. This is a key thing.
All other arguments should be based on this.

> What would be a meaningful timeout for the 10 Sec test running against a mutant?

Again, I think the question is incorrect. Mutation testing framework (MTF) works with a project test suite (many tests), not with particular test. And the value of `timeout` property is passed to the Symfony Process that runs _all_ tests.

> As an example: My current project has 1,566 tests and executes in 3 Minutes, which means that infection can run for 78 hours.

Practically this is not true, because MTF has the following optimizations (may be it will clarify timeout property as well):

1. For each mutant MTF runs *only those tests that cover mutated line of code*. And it runs it with `--stop-on-failure` option. It means that only 3 (just an example) tests will be run of 1,566 for particular mutant.
2. Also, MTF runs the fastest tests first. So if in your example mutated line is covered by 2 tests, one is 40s (name it `A`), one is 10s (name it `B`), the first executed test will be `B`.

It also means that as soon as one of the tests fails, this Mutant is considered Killed. No need to wait `timeout` seconds for this process.

> But having the same timeout in the whole run, makes this currently impossible :(

For your example, if you set `timeout=10s` and run MTF for mutant that is covered by test `A` (40s), then it will be timeouted (===killed), which is not true! This is a false-positve result, because test `A` potentially can not kill the mutant.
In other words, timeout should be = or > than the whole test suite takes, otherwise you *can* get false-positives.

> What are the quality requirements for PR's?

We just follow some rules that are outlined in `.php_cs.dist` file in the project root (rules for PHP CS Fixer). In any case build will fail if you do something wrong ;)

------

Does it make sense?

**Comment by theofidry** [comment]
Since #41 has been merged is there still something to address here or can this issue be closed?
